graphql-protocol/src/main/java/com/linecorp/armeria/server/graphql/protocol/MultipartVariableMapper.java

```diff
@@ -0,0 +1,147 @@
+         return location.put(target, path);
+      }
+
+      @Override
+      public Object recurse(Map<String, Object> location, String target) {
+         return location.get(target);
+      }
+   };
+   private static final Mapper<List<Object>> LIST_MAPPER = new Mapper<List<Object>>() {
+      @Override
+      public Object set(List<Object> location, String target, Path path) {
+         final Integer value = parseUnsignedInt(target);
+         if (value == null) {
+            return location;
+         }
+         if (value >= location.size()) {
+            return location;
+         }
+         return location.set(value, path);
+      }
+
+      @Override
+      public Object recurse(List<Object> location, String target) {
+         final Integer value = parseUnsignedInt(target);
+         if (value == null) {
+            return null;
+         }
+         if (value >= location.size()) {
+            return null;
+         }
+         return location.get(value);
+      }
+   };
+
+   static void mapVariable(String objectPath, Map<String, Object> variables, Path path) {
+      final String[] segments = PERIOD.split(objectPath);
+      if (segments.length < 2) {
+         throw new IllegalArgumentException("object-path in map must have at least two segments");
+      }
+      if (!"variables".equals(segments[0])) {
+         throw new IllegalArgumentException("can only map into variables");
+      }
+
+      Object currentLocation = variables;
+      for (int i = 1; i < segments.length; i++) {
+         final String segmentName = segments[i];
+         if (i == segments.length - 1) {
+            if (currentLocation instanceof Map) {
+               if (MAP_MAPPER.set((Map<String, Object>) currentLocation, segmentName, path) != null) {
+                  throw new IllegalArgumentException("expected null value when mapping " + objectPath);
+               }
+            } else {
+               if (LIST_MAPPER.set((List<Object>) currentLocation, segmentName, path) != null) {
+                  throw new IllegalArgumentException("expected null value when mapping " + objectPath);
+               }
+            }
+         } else {
+            if (currentLocation instanceof Map) {
+               currentLocation = MAP_MAPPER.recurse((Map<String, Object>) currentLocation, segmentName);
+            } else {
+               currentLocation = LIST_MAPPER.recurse((List<Object>) currentLocation, segmentName);
+            }
+            if (currentLocation == null) {
+               throw new IllegalArgumentException(
+                       "found null intermediate value when trying to map " + objectPath);
+            }
+         }
+      }
+   }
+
+   interface Mapper<T> {
+      @Nullable
+      Object set(T location, String target, Path path);
+
+      @Nullable
+      Object recurse(T location, String target);
+   }
+
+   @Nullable
+   private static Integer parseUnsignedInt(String value) {
```

Comment:
How about returning `-1` when failed to parse, so we allocate fewer objects?

Reply:
That's a very good idea! 😎